ironfish/src/migrations/migrator.ts

```diff
@@ -0,0 +1,165 @@
+    this.node = options.node
+    this.logger = options.logger
+
+    this.migrations = MIGRATIONS.map((m) => new m().init(options.node.files)).sort(
+      (a, b) => a.id - b.id,
+    )
+  }
+
+  /**
+   * Returns true if any database is at version 0
+   */
+  async isInitial(): Promise<boolean> {
+    for (const migration of this.migrations) {
+      if (await this.isEmpty(migration)) {
+        return true
+      }
+    }
+    return false
+  }
+
+  /**
+   * Returns true if the migration database is at version 0
+   */
+  async isEmpty(migration: Migration): Promise<boolean> {
+    const db = await migration.prepare(this.node)
+
+    try {
+      await db.open()
+      const version = await db.getVersion()
+      return version === 0
+    } finally {
+      await db.close()
+    }
+  }
+
+  async isApplied(migration: Migration): Promise<boolean> {
+    const db = await migration.prepare(this.node)
+
+    try {
+      await db.open()
+      const version = await db.getVersion()
+      return version >= migration.id
+    } finally {
+      await db.close()
+    }
+  }
+
+  async revert(): Promise<void> {
+    const migrations = Array.from(this.migrations).reverse()
+
+    for (const migration of migrations) {
+      const applied = await this.isApplied(migration)
+
+      if (applied) {
+        this.logger.info(`Reverting ${migration.name}`)
+        const db = await migration.prepare(this.node)
+
+        try {
+          await db.open()
+
+          await db.transaction(async (tx) => {
+            await migration.backward(this.node, db, tx)
+            await db.putVersion(migration.id - 1, tx)
+          })
+        } finally {
+          await db.close()
+        }
+
+        return
+      }
+    }
+  }
+
+  async migrate(options?: { quiet?: boolean; quietNoop?: boolean }): Promise<void> {
+    const logger = options?.quiet ? null : this.logger
+    const writeOut = options?.quiet ? null : (t: string) => process.stdout.write(t)
+
+    const status = new Array<[Migration, boolean]>()
+    for (const migration of this.migrations) {
+      const applied = await this.isApplied(migration)
+      status.push([migration, applied])
+    }
+
+    const unapplied = status.filter(([, applied]) => !applied).map(([migration]) => migration)
+
+    if (unapplied.length === 0) {
+      if (!options?.quietNoop) {
+        logger?.info(`All ${this.migrations.length} migrations applied.`)
+      }
+      return
+    }
+
+    logger?.info(`Running ${unapplied.length} migrations:`)
+
+    for (const migration of unapplied) {
+      writeOut?.(`  Applying ${migration.name}...`)
+
+      const db = await migration.prepare(this.node)
+
+      try {
+        await db.open()
+
+        await db.transaction(async (tx) => {
+          await migration.forward(this.node, db, tx)
+          await db.putVersion(migration.id, tx)
+        })
+      } catch (e) {
+        writeOut?.(` ERROR\n`)
+        this.logger.error(ErrorUtils.renderError(e, true))
```

Comment:
Should this be:

```suggestion
        logger?.error(ErrorUtils.renderError(e, true))
```

Or will we always log it out because it's an error?

Reply:
Always log during a crash.